kernel/arch/dreamcast/util/screenshot.c

```diff
@@ -24,53 +27,88 @@ This will now work with any of the supported video modes.
         return -1;
     }
 
+    /* Open output file */
+    f = fs_open(destfn, O_WRONLY | O_TRUNC);
+
+    if(!f) {
+        dbglog(DBG_ERROR, "vid_screen_shot: can't open output file '%s'\n", destfn);
+        free(buffer);
+        return -1;
+    }
+
     /* Disable interrupts */
     save = irq_disable();
 
-    /* Write out each pixel as 24 bits */
+    /* Write out each pixel as 24-bits */
     switch(vid_mode->pm) {
         case(PM_RGB555): {
-            for(i = 0; i < numpix; i++) {
-                pixel = vram_s[i];
-                r = (((pixel >> 10) & 0x1f) << 3);
-                g = (((pixel >>  5) & 0x1f) << 3);
-                b = (((pixel >>  0) & 0x1f) << 3);
-                buffer[i * 3 + 0] = r;
-                buffer[i * 3 + 1] = g;
-                buffer[i * 3 + 2] = b;
+            /* Process two 16-bit pixels at a time */
+            for(i = 0; i < numpix/2; i++) {
+                pixel = vram_l[i];
+                pixel1 = pixel & 0xFFFF;
+                pixel2 = pixel >> 16;
+
+                /* Process the first pixel */
+                r = (((pixel1 >> 10) & 0x1f) << 3);
+                g = (((pixel1 >> 5) & 0x1f) << 3);
+                b = (((pixel1 >> 0) & 0x1f) << 3);
+                buffer[i * 6 + 0] = r;
+                buffer[i * 6 + 1] = g;
+                buffer[i * 6 + 2] = b;
+
+                /* Process the second pixel */
+                r = (((pixel2 >> 10) & 0x1f) << 3);
+                g = (((pixel2 >> 5) & 0x1f) << 3);
+                b = (((pixel2 >> 0) & 0x1f) << 3);
+                buffer[i * 6 + 3] = r;
+                buffer[i * 6 + 4] = g;
+                buffer[i * 6 + 5] = b;
             }
 
             break;
         }
         case(PM_RGB565): {
-            for(i = 0; i < numpix; i++) {
-                pixel = vram_s[i];
-                r = (((pixel >> 11) & 0x1f) << 3);
-                g = (((pixel >>  5) & 0x3f) << 2);
-                b = (((pixel >>  0) & 0x1f) << 3);
-                buffer[i * 3 + 0] = r;
-                buffer[i * 3 + 1] = g;
-                buffer[i * 3 + 2] = b;
+            /* Process two 16-bit pixels at a time */
+            for(i = 0; i < numpix/2; i++) {
+                pixel = vram_l[i];
+                pixel1 = pixel & 0xFFFF;
+                pixel2 = pixel >> 16;
+
+                /* Process the first pixel */
+                r = (((pixel1 >> 11) & 0x1f) << 3);
+                g = (((pixel1 >> 5) & 0x3f) << 2);
+                b = (((pixel1 >> 0) & 0x1f) << 3);
+                buffer[i * 6 + 0] = r;
+                buffer[i * 6 + 1] = g;
+                buffer[i * 6 + 2] = b;
+
+                /* Process the second pixel */
+                r = (((pixel2 >> 11) & 0x1f) << 3);
+                g = (((pixel2 >> 5) & 0x3f) << 2);
+                b = (((pixel2 >> 0) & 0x1f) << 3);
+                buffer[i * 6 + 3] = r;
+                buffer[i * 6 + 4] = g;
+                buffer[i * 6 + 5] = b;
```

Comment:
You can add other pixel modes too:
https://github.com/DC-SWAT/DreamShell/blob/master/firmware/isoldr/loader/utils.c?ts=4#L638

Reply:
Added support for PM_RGB888P